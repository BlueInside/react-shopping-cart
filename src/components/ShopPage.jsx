import LoadingSpinner from './styles/LoadingSpinner.styled';
import { ProductsDisplay } from './styles/ShopPage.styled';
import { StyledFilterDropdown } from './styles/ShopPage.styled';

import useDataFetching from '../hooks/useDataFetching';
import ProductCard from './ProductCard';
import { useState } from 'react';
import { Outlet } from 'react-router-dom';

function ShopPage() {
  const [sortOption, setSortOption] = useState('default');

  let { data, loading, error } = useDataFetching();

  function sortData(option) {
    const sortedData = [...data];

    if (option === 'az') {
      sortedData.sort((a, b) => a.title.localeCompare(b.title));
    } else if (option === 'za') {
      sortedData.sort((a, b) => b.title.localeCompare(a.title));
    }

    if (option === 'lowHigh') {
      sortedData.sort((a, b) => a.price - b.price);
    } else if (option === 'highLow') {
      sortedData.sort((a, b) => b.price - a.price);
    }

    return sortedData;
  }

  if (loading) {
    return <LoadingSpinner />;
  }
  if (error) {
    return <div>Oops! Something went wrong. Please try again later.</div>;
  }
  return (
    <>
      <Outlet />
      <StyledFilterDropdown setSortOption={setSortOption} />
      <ProductsDisplay>
        {sortData(sortOption).map((product) => (
          <ProductCard
            product={{ ...product }}
            key={product.id}
            image={product.image}
            price={product.price}
            title={product.title}
            description={product.description}
            rating={product.rating}
            category={product.category}
          />
        ))}
      </ProductsDisplay>
    </>
  );
}

export default ShopPage;
