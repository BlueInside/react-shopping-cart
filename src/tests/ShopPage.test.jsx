import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import ShopPage from '../components/ShopPage';
describe('ShopPage component', () => {
  it('renders ShopPage correctly', () => {
    const { container } = render(<ShopPage />);
    expect(container).toMatchSnapshot();
  });
});
