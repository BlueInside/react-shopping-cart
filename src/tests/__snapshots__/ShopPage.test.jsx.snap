// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShopPage component > renders ShopPage correctly 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
